snuba/query/snql/__init__.py

```diff
@@ -1,23 +1,187 @@
+    HighPriTuple,
+    LowPriArithmetic,
+    HighPriArithmetic,
+)
 
+snql_grammar = Grammar(
+    fr"""
+    query_exp             = match_clause? where_clause? collect_clause? having_clause? order_by_clause?
+    match_clause          = space* "MATCH" space* "(" clause ")" space*
+    where_clause          = space* "WHERE" clause space*
+    collect_clause        = space* "COLLECT" collect_list space*
+    order_by_clause       = space* "ORDER BY" collect_list ("ASC"/"DESC") space*
 
-grammar = Grammar(
-    r"""
-    query_exp = match_clause where_clause? collect_clause? having_clause? order_by_clause?
-    match_clause = space* "MATCH" "(" clause ")" space*
-    where_clause = space* "WHERE" clause space*
-    collect_clause = space* "COLLECT" clause "BY" clause space*
-    having_clause = space* "HAVING" clause space*
-    order_by_clause = space* "ORDER BY" clause ("ASC"/"DESC") space*
+    collect_list          = collect_columns* (low_pri_arithmetic)
+    collect_columns       = low_pri_arithmetic space* comma space*
+
+    having_clause         = space* "HAVING" clause space*
+    clause                = space* ~r"[-=><\w]+" space*
+
+    low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
+    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    low_pri_tuple         = low_pri_op high_pri_arithmetic
+    high_pri_tuple        = high_pri_op arithmetic_term
+    arithmetic_term       = (space*) (function_call / numeric_literal / column_name) (space*)
+    low_pri_op            = "+" / "-"
+    high_pri_op           = "/" / "*"
+    param_expression      = low_pri_arithmetic / quoted_literal
+    parameters_list       = parameter* (param_expression)
+    parameter             = param_expression space* comma space*
+    function_call         = function_name open_paren parameters_list? close_paren (open_paren parameters_list? close_paren)?
+    simple_term           = quoted_literal / numeric_literal / column_name
+    literal               = ~r"[a-zA-Z0-9_\.:-]+"
+    quoted_literal        = "'" string_literal "'"
+    string_literal        = ~r"[a-zA-Z0-9_\.\+\*\/:-]*"
+    numeric_literal       = ~r"-?[0-9]+(\.[0-9]+)?(e[\+\-][0-9]+)?"
+    column_name           = ~r"[a-zA-Z_][a-zA-Z0-9_\.]*"
+    function_name         = ~r"[a-zA-Z_][a-zA-Z0-9_]*"
+    open_paren            = "("
+    close_paren           = ")"
+    space                 = " "
+    comma                 = ","
 
-    clause = space* ~r"[-=><\w]+" space*
-    space = " "
 """
 )
 
 
+class SnQLVisitor(NodeVisitor):
+    """
+    Builds Snuba AST expressions from the Parsimonious parse tree.
+    """
+
+    def visit_query_exp(self, node, visited_children):
+        _, _, collect, _, _ = visited_children
+
+        return Query({}, None, collect, None,)
+
+    def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> str:
+        return ExpressionVisitor.visit_function_name(self, node, visited_children)
```

Comment:
This is not supposed to work. You should not pass self into an instance method of a class.  Python takes care of that. 
If you want to call a method without having an instance of the class you have two alternatives.
1) Just don't create the class. You do not need to define the ExpressionVisitor class, all the methods are class methods (do not rely on the instance), so why not just defining them as functions of the expression_visitor module, and here you import directly the functions.
2) define them as class methods of the class ExpressionVisitor. Which means you decorate the method with `@classmethod` and, from here you do not pass self when calling the method. 

I would suggest option 1 since you do not have any method in ExpressionVisitor that depends on the state of the class, so there is no reason for having the class.

Reply:
I applied option 1. It seems to me that the issue I was facing earlier still exists: for some reason, there are 2 identical (correct) Query objects being produced.